### PR TITLE
shmem_calloc needs to allocate count*size not size

### PR DESCRIPTION
**Pull request by jeffhammond**
resolves issue #122

Signed-off-by: Jeff Hammond <jeff.science@gmail.com>

### CODE REVIEW

**Comment by jeffhammond** [comment]
Ah the change to give allocation barrier semantics is an OpenSHMEM thing that wasn't in SHMEM before. I'll add a barrier at the end.

**Comment by jeffhammond** [comment]
IIRC, shmem_barrier_all = MPI_Barrier whereas shmem_sync_all = MPI_Win_flush_all + MPI_Barrier, so the former is cheaper, but I'll have to look to be sure.

I don't see a strong reason to elide the synchronize on count=0.  I don't want to reward users for calling allocation functions that don't do anything :-)

**Comment by jeffhammond** [comment]
@dalcinl i implemented it differently and now it should be both correct and efficient

File: src/shmem/mem.c [hunk 1]
```diff
@@ -41,6 +41,10 @@ void *shmem_calloc(size_t count, size_t size)
 {
     void *ptr = NULL;
 
+    /* When count or size is 0, the shmem_calloc routine returns without performing a barrier.
+     * Otherwise, this routine calls a procedure that is semantically equivalent to shmem_barrier_all on exit. */
+    if ( (count==0) || (size==0) ) { return ptr; }
+
```
Comment (hzhou) [inline src/shmem/mem.c:47]: Is the behavior of a barrier still required by the specification?

Comment (hzhou) [reply src/shmem/mem.c:47]: Nevermind. Just read the discussion. :) Could you make it clear that the comment is quoting v1.5 specification?

**Review by hzhou — CHANGES_REQUESTED** [review]
The title needs an update to reflect the new changes. Probably you should have 3 commits, the 1st commit simply correct the size.

**Comment by hzhou** [comment]
The whitespace checker demands:
```
-    if ((count == 0) || (size == 0)) {
-        return ptr;
-    }
+    if ( (count==0) || (size==0) ) { return ptr; }
```

File: src/shmem/mem.c (hunk 1 above)
Comment (jeffhammond) [reply src/shmem/mem.c:48]: done f9f940fd8344994e6c34397093ba9af16d5cb335

**Comment by jeffhammond** [comment]
> The title needs an update to reflect the new changes. Probably you should have 3 commits, the 1st commit simply correct the size.

I do not understand this.  Change the title of what?

As for the commits, merely fixing the count*size issue requires either a potentially incorrect implementation or adding two barriers.  I think it is reasonable to just do it right in one shot.

**Comment by hzhou** [comment]
> > The title needs an update to reflect the new changes. Probably you should have 3 commits, the 1st commit simply correct the size.
> 
> I do not understand this. Change the title of what?
> 
> As for the commits, merely fixing the count*size issue requires either a potentially incorrect implementation or adding two barriers. I think it is reasonable to just do it right in one shot.

I guess my peeve is the commit message --
```
push calloc impl down into OSHMPI layer

malloc+memset impl of calloc required two calls to barrier_all.
pushing calloc down to OSHMPI layer eliminates the unnecessary barrier_all.
a side benefit is that debugging output now exposes calloc calls directly.
```
has nothing referring to the title. Maybe just add a sentence explaining the original reason for the change?

**Comment by hzhou** [comment]
test:oshmpi/warnings

**Comment by jeffhammond** [comment]
there is now one commit, with all the relevant information.  i do not know how to edit commit messages for any but the most recent message, which is why i did it this way.

**Comment by hzhou** [comment]
> there is now one commit, with all the relevant information. i do not know how to edit commit messages for any but the most recent message, which is why i did it this way.

I understand 😄 Would you like to learn (and rid of the excuse from now on) -- https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history?

**Comment by hzhou** [comment]
@jeffhammond FYI, I just cleaned up the warnings in https://github.com/pmodels/oshmpi/pull/133. We'll ask you for a rebase once that is merged.

**Pull request merged**